pkg/common-controller/snapshot_controller.go

```diff
@@ -425,27 +438,38 @@ func (ctrl *csiSnapshotCommonController) syncUnreadySnapshot(snapshot *crdv1.Vol
 			// update snapshot status failed
 			klog.V(4).Infof("failed to update snapshot %s status: %v", utils.SnapshotKey(snapshot), err)
 			ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotStatusUpdateFailed", fmt.Sprintf("Snapshot status update failed, %v", err))
+			ctrl.RecordSnapshotMetrics(snapshot, createSnapshotOperationName, preProvisionedSnapshotType, err)
 			return err
 		}
+
+		ctrl.RecordSnapshotMetrics(snapshot, createSnapshotOperationName, preProvisionedSnapshotType, nil)
 		return nil
 	}
+
 	// snapshot.Spec.Source.VolumeSnapshotContentName == nil - dynamically creating snapshot
+	ctrl.StartMetricsOperation(snapshot, snapshottingOperationName, dynamicSnapshotType)
```

Comment:
I am not sure whether the timer should be started here.
line 451 to 477 is really to handle the situations when the controller crashed/restarted after a content has been created however binding has not been successfully executed.
Or is it your desire to only recode error counts? then at line 475 it should clean the operation, otherwise it would be memory leak?

Reply:
I am not sure either. I thought this code block inside of syncUnreadySnapshot was the extent of pre-provisioned snapshot handling:
https://github.com/kubernetes-csi/external-snapshotter/pull/409/files#diff-1a3473f9016f7abcf5007414ff7141e39032aaf7aba079eeef262d04167208a2R423-R433

Is there another area where we handle this case?